### PR TITLE
Introduce PCI devices

### PR DESCRIPTION
**Pull request by bsatzger**
This PR adds a model representing PCI devices. A PCI device always belongs to a single VM host and may be bound to a single VM. In addition, it adds the ability to `HostNexus` to learn about PCI devices. For now, we only consider NVIDIA devices for which IOMMU is enabled.

### CODE REVIEW

File: migrate/20240507_pci_device.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+      column :id, :uuid, primary_key: true
+      column :slot, :text, null: false
+      column :device_class, :text, null: false
+      column :vendor, :text, null: false
+      column :device, :text, null: false
+      column :numa_node, :Integer, null: true
+      column :iommu_group, :Integer, null: false
+      column :enabled, :bool, null: false, default: true
+      foreign_key :vm_host_id, :vm_host, type: :uuid, null: false
+      foreign_key :vm_id, :vm, type: :uuid, null: true
```
Comment (bsatzger) [inline migrate/20240507_pci_device.rb:15]: We will probably not have many PCI devices, initially, but we might want to add an index at some point on `vm_id`.

File: model/pci_device.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+require_relative "../model"
+
+class PciDevice < Sequel::Model
```
Comment (enescakir) [inline model/pci_device.rb:5]: The new model and migration should be in separate commits for zero downtime deployment. Otherwise, the app won't start in production until you run the migration. We initially froze all models.

```
➜  ubicloud git:(ben/learn_gpu) RACK_ENV=production bundle exec rackup
bundler: failed to load command: rackup (/Users/enescakir/.asdf/installs/ruby/3.2.3/bin/rackup)
/Users/enescakir/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/sequel-5.79.0/lib/sequel/plugins/require_valid_schema.rb:58:in `get_db_schema_array': Not able to parse schema for model: PciDevice, table: "pci_device" (Sequel::Error)
        from /Users/enescakir/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/sequel-5.79.0/lib/sequel/model/base.rb:792:in `get_db_schema'
        from /Users/enescakir/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/sequel-5.79.0/lib/sequel/model/base.rb:606:in `set_dataset'
        from /Users/enescakir/.asdf/installs/ruby/3.2.3/lib/ruby/gems/3.2.0/gems/sequel-5.79.0/lib/sequel/model/plugins.rb:52:in `block in after_set_data
```

File: model/pci_device.rb [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+  include ResourceMethods
+
+  many_to_one :vm_host
+  many_to_one :vm, key: :vm_id, class: :PciDevice
```
Comment (enescakir) [inline model/pci_device.rb:9]:
```suggestion
  many_to_one :vm
```

File: model/vm.rb [hunk 4]
```diff
@@ -11,6 +11,7 @@ class Vm < Sequel::Model
   one_to_one :assigned_vm_address, key: :dst_vm_id, class: :AssignedVmAddress
   one_to_many :vm_storage_volumes, key: :vm_id, order: Sequel.desc(:boot)
   one_to_one :active_billing_record, class: :BillingRecord, key: :resource_id do |ds| ds.active end
+  one_to_many :pci_device, key: :vm_id, class: :PciDevice
```
Comment (enescakir) [inline model/vm.rb:14]: Isn't it a `one_to_one` relation? If a vm can have multiple PciDevice, IMHO, we need to rename is as `:pci_devices`.

File: model/vm_host.rb [hunk 5]
```diff
@@ -12,6 +12,7 @@ class VmHost < Sequel::Model
   one_to_many :assigned_host_addresses, key: :host_id, class: :AssignedHostAddress
   one_to_many :spdk_installations, key: :vm_host_id
   one_to_many :storage_devices, key: :vm_host_id
+  one_to_many :pci_device, key: :vm_host_id
```
Comment (enescakir) [inline model/vm_host.rb:15]:
```suggestion
  one_to_many :pci_devices, key: :vm_host_id
```

File: migrate/20240507_pci_device.rb (hunk 1 above)
Comment (bsatzger) [reply migrate/20240507_pci_device.rb:15]: Added an index.

File: model/vm.rb (hunk 4 above)
Comment (bsatzger) [reply model/vm.rb]: A vm can have multiple PciDevices, renamed to `:pci_devices`

File: model/pci_device.rb (hunk 2 above)
Comment (bsatzger) [reply model/pci_device.rb:5]: thank you!

File: prog/learn_pci.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        fail "BUG: lspci parse failed" unless REQUIRED_KEYS.all? { |s| dev_h.key? s }
+        next unless dev_h.key? "IOMMUGroup"
+        out << PciDeviceRecord.new(dev_h["Slot"], dev_h["Class"], dev_h["Vendor"], dev_h["Device"], dev_h["NUMANode"], dev_h["IOMMUGroup"])
+      end
+      out.freeze
+    end
+  end
+
+  def make_model_instances
+    PciDeviceRecord.parse_all(sshable.cmd("/usr/bin/lspci -vnmm -d 10de::")).map do |rec|
```
Comment (furkansahin) [inline prog/learn_pci.rb:21]: Note to future if someone comes here: 10de is the vendor id for nvidia

**Review by furkansahin — APPROVED** [review]

**Review by enescakir — APPROVED** [review]
